File: frontend/src/api/workflows.ts
import { apiClient } from "@/lib/api-client";
import type { DataResponse, PaginatedResponse } from "@/types/api";

export type WorkflowType = "phl" | "sra";
export type WorkflowStatus = "draft" | "submitted" | "approved" | "rejected";

export interface Workflow {
  id: string;
  organization_id: string;
  created_by: string;
  type: WorkflowType;
  status: WorkflowStatus;
  title: string;
  data: Record<string, unknown>;
  conversation_id: string | null;
  submitted_at: string | null;
  approved_at: string | null;
  approved_by: string | null;
  risk_entry_id: string | null;
  created_at: string;
  updated_at: string;
}

export interface CreateWorkflowPayload {
  type: WorkflowType;
  title?: string;
  data?: Record<string, unknown>;
  conversation_id?: string | null;
}

export interface UpdateWorkflowPayload {
  title?: string;
  data?: Record<string, unknown>;
}

export async function createWorkflow(
  payload: CreateWorkflowPayload,
): Promise<Workflow> {
  const response = await apiClient.post<DataResponse<Workflow>>(
    "/workflows",
    payload,
  );
  return response.data.data;
}

export async function getWorkflow(id: string): Promise<Workflow> {
  const response = await apiClient.get<DataResponse<Workflow>>(
    `/workflows/${id}`,
  );
  return response.data.data;
}

export async function listWorkflows(params?: {
  type?: WorkflowType;
  status?: WorkflowStatus;
  skip?: number;
  limit?: number;
}): Promise<PaginatedResponse<Workflow>> {
  const response = await apiClient.get<PaginatedResponse<Workflow>>(
    "/workflows",
    { params },
  );
  return response.data;
}

export async function updateWorkflow(
  id: string,
  payload: UpdateWorkflowPayload,
): Promise<Workflow> {
  const response = await apiClient.patch<DataResponse<Workflow>>(
    `/workflows/${id}`,
    payload,
  );
  return response.data.data;
}

export async function submitWorkflow(id: string): Promise<Workflow> {
  const response = await apiClient.post<DataResponse<Workflow>>(
    `/workflows/${id}/submit`,
  );
  return response.data.data;
}

export async function approveWorkflow(
  id: string,
  approve: boolean,
  notes?: string,
): Promise<Workflow> {
  const response = await apiClient.post<DataResponse<Workflow>>(
    `/workflows/${id}/approve`,
    { approve, notes },
  );
  return response.data.data;
}

export async function deleteWorkflow(id: string): Promise<void> {
  await apiClient.delete(`/workflows/${id}`);
}

File: backend/app/api/v1/workflows.py
import uuid

from fastapi import APIRouter, Depends, Query
from sqlalchemy.ext.asyncio import AsyncSession

from app.core.database import get_db
from app.core.deps import (
    get_audit_logger,
    get_current_organization,
    require_analyst_or_above,
    require_org_role,
)
from app.models.organization import Organization
from app.models.organization_membership import MembershipRole
from app.models.user import User
from app.models.workflow import WorkflowStatus, WorkflowType
from app.schemas.common import DataResponse, MetaResponse, PaginatedMeta, PaginatedResponse
from app.schemas.workflow import (
    ApproveWorkflowRequest,
    CreateWorkflowRequest,
    UpdateWorkflowRequest,
    WorkflowResponse,
)
from app.services.audit import AuditLogger
from app.services.workflow import WorkflowService

router = APIRouter(prefix="/workflows", tags=["workflows"])


def _get_workflow_service(db: AsyncSession = Depends(get_db)) -> WorkflowService:
    return WorkflowService(db=db)


@router.post("", response_model=DataResponse[WorkflowResponse], status_code=201)
async def create_workflow(
    payload: CreateWorkflowRequest,
    current_user: User = Depends(require_analyst_or_above),
    organization: Organization = Depends(get_current_organization),
    service: WorkflowService = Depends(_get_workflow_service),
    audit: AuditLogger = Depends(get_audit_logger),
) -> DataResponse[WorkflowResponse]:
    workflow = await service.create(payload, current_user.id, organization.id)
    await audit.log(
        action=f"workflow.{payload.type.value}.created",
        user=current_user,
        resource_type="workflow",
        resource_id=str(workflow.id),
        organization_id=organization.id,
    )
    return DataResponse(
        data=WorkflowResponse.model_validate(workflow),
        meta=MetaResponse(request_id=str(workflow.id)),
    )


@router.get("", response_model=PaginatedResponse[WorkflowResponse])
async def list_workflows(
    skip: int = Query(0, ge=0),
    limit: int = Query(50, ge=1, le=100),
    type: WorkflowType | None = Query(None),
    status: WorkflowStatus | None = Query(None),
    current_user: User = Depends(require_analyst_or_above),
    organization: Organization = Depends(get_current_organization),
    service: WorkflowService = Depends(_get_workflow_service),
) -> PaginatedResponse[WorkflowResponse]:
    workflows, total = await service.list_for_org(
        organization_id=organization.id,
        type_filter=type,
        status_filter=status,
        skip=skip,
        limit=limit,
    )
    items = [WorkflowResponse.model_validate(w) for w in workflows]
    total_pages = (total + limit - 1) // limit if total else 0
    return PaginatedResponse(
        data=items,
        meta=PaginatedMeta(
            request_id="",
            total=total,
            page=(skip // limit) + 1,
            page_size=limit,
            total_pages=total_pages,
        ),
    )


@router.get("/{workflow_id}", response_model=DataResponse[WorkflowResponse])
async def get_workflow(
    workflow_id: uuid.UUID,
    current_user: User = Depends(require_analyst_or_above),
    organization: Organization = Depends(get_current_organization),
    service: WorkflowService = Depends(_get_workflow_service),
) -> DataResponse[WorkflowResponse]:
    workflow = await service.get(workflow_id, organization.id)
    return DataResponse(
        data=WorkflowResponse.model_validate(workflow),
        meta=MetaResponse(request_id=str(workflow_id)),
    )


@router.patch("/{workflow_id}", response_model=DataResponse[WorkflowResponse])
async def update_workflow(
    workflow_id: uuid.UUID,
    payload: UpdateWorkflowRequest,
    current_user: User = Depends(require_analyst_or_above),
    organization: Organization = Depends(get_current_organization),
    service: WorkflowService = Depends(_get_workflow_service),
    audit: AuditLogger = Depends(get_audit_logger),
) -> DataResponse[WorkflowResponse]:
    workflow = await service.update(workflow_id, organization.id, payload)
    await audit.log(
        action=f"workflow.{workflow.type.value}.updated",
        user=current_user,
        resource_type="workflow",
        resource_id=str(workflow_id),
        organization_id=organization.id,
    )
    return DataResponse(
        data=WorkflowResponse.model_validate(workflow),
        meta=MetaResponse(request_id=str(workflow_id)),
    )


@router.post("/{workflow_id}/submit", response_model=DataResponse[WorkflowResponse])
async def submit_workflow(
    workflow_id: uuid.UUID,
    current_user: User = Depends(require_analyst_or_above),
    organization: Organization = Depends(get_current_organization),
    service: WorkflowService = Depends(_get_workflow_service),
    audit: AuditLogger = Depends(get_audit_logger),
) -> DataResponse[WorkflowResponse]:
    workflow = await service.submit(workflow_id, organization.id)
    await audit.log(
        action=f"workflow.{workflow.type.value}.submitted",
        user=current_user,
        resource_type="workflow",
        resource_id=str(workflow_id),
        organization_id=organization.id,
    )
    return DataResponse(
        data=WorkflowResponse.model_validate(workflow),
        meta=MetaResponse(request_id=str(workflow_id)),
    )


@router.post("/{workflow_id}/approve", response_model=DataResponse[WorkflowResponse])
async def approve_workflow(
    workflow_id: uuid.UUID,
    payload: ApproveWorkflowRequest,
    current_user: User = Depends(require_org_role(MembershipRole.ORG_ADMIN)),
    organization: Organization = Depends(get_current_organization),
    service: WorkflowService = Depends(_get_workflow_service),
    audit: AuditLogger = Depends(get_audit_logger),
) -> DataResponse[WorkflowResponse]:
    workflow = await service.approve(workflow_id, organization.id, current_user.id, payload.approve)
    await audit.log(
        action=f"workflow.{workflow.type.value}.{'approved' if payload.approve else 'rejected'}",
        user=current_user,
        resource_type="workflow",
        resource_id=str(workflow_id),
        organization_id=organization.id,
        metadata={"notes": payload.notes} if payload.notes else None,
    )
    return DataResponse(
        data=WorkflowResponse.model_validate(workflow),
        meta=MetaResponse(request_id=str(workflow_id)),
    )


@router.delete("/{workflow_id}", status_code=204)
async def delete_workflow(
    workflow_id: uuid.UUID,
    current_user: User = Depends(require_analyst_or_above),
    organization: Organization = Depends(get_current_organization),
    service: WorkflowService = Depends(_get_workflow_service),
    audit: AuditLogger = Depends(get_audit_logger),
) -> None:
    await service.delete(workflow_id, organization.id)
    await audit.log(
        action="workflow.deleted",
        user=current_user,
        resource_type="workflow",
        resource_id=str(workflow_id),
        organization_id=organization.id,
    )
